docs/RFCS/20171003_inverted_indexes.md

```diff
@@ -0,0 +1,570 @@
+example above would produce the following index keys:
+
+
+    1/2/arr/3/pk1
+    1/2/arr/a/pk1
+    1/2/arr/arr/4/pk1
+    1/2/arr/arr/5/pk1
+    1/2/arr/arr/false/pk1
+    1/2/arr/arr/true/pk1
+    1/2/arr/foo/bar/pk1
+
+We don’t actually distinguish arrays with an index or name, just a level of
+nesting. To understand why, consider the following query:
+
+    SELECT * from t where jsonCol @> [[false]];
+
+This should return the JSON above because there is indeed a value `false`
+nested two deep in an array. However, if we included the indices of the array
+values, we wouldn’t be able to disambiguate between the inner array `[4,5]` and
+`[false, true]`. If we just add a prefix for array containment, we don’t have
+this problem.
+
+## Detailed design: operators, index selection, index management
+
+**Supported Operators**
+
+Postgres provides the following indexable operators on JSON:
+
+| @> | Does the left JSON value contain the right JSON path/value entries at the top level? |
+| -- | ------------------------------------------------------------------------------------ |
+| ?  | Does the string exist as a top-level key within the JSON value?                      |
+| ?& | Do all of these array strings exist as top-level keys?                               |
+| ?| | Do any of these array strings exist as top-level keys?                               |
+
+Note that, unlike Postgres’ `json_path_ops` index, the proposed index will
+support all of the operators above, not just the first one. The reason for this
+extra support is that the proposed encoding is lossless with respect to the
+structure of the input JSON, unlike Postgres’ lossy hash encoding; searching
+for a top-level key is as simple as performing a range scan. For example, given
+the column value `{a: {c: 3}}` and the resultant index key `…/a/c/3/`,
+completing the query `col ? a` merely requires checking to see if any rows are
+available in the keyspan `/a → /b`.
+
+Even though supporting this feature isn't in scope for initial implementation,
+the proposed index key encoding permits range queries and partial path matches
+that are not even expressible by Postgres’s `@>` operator semantics. For
+example, the query `SELECT * FROM j WHERE col->a->>x > 3`, though inexpressible
+via Postgres’ indexable JSON operators, is easily expressed via a scan over
+keyspan `/a/3 -> /b`.
+
+**Index Creation**
+The syntax of creating an inverted index on a `JSONB` column type should match
+Postgres, for compatibility with existing Postgres applications.
+
+Therefore, we must support the following syntax even though we don’t support
+Postgres’ GIN indexes in general:
+
+    CREATE INDEX <optional name> ON <table> USING GIN(<column> <opt_ops_specifier>)
+
+where `<opt_ops_specifier>` can be only `jsonb_path_ops`, since that’s the
+index mode we’ve chosen to support.
+
+For ergonomics, we might want to also support a syntax like the following:
```

Comment:
I'm not too keen on adding this for ergonomics. The USING GIN style seems okay for now.

Reply:
@vivekmenezes can you elaborate why you are not keen on adding the extra syntax?

In my opinion, we should support both and *only* document the one that is CockroachDB-specific. I do not think we should document the "GIN" syntax until we support more of pg's feature, for the reasons that Peter explain. However perhaps I can be convinced otherwise.